src/cipher/mod.rs

```diff
@@ -0,0 +1,359 @@
+/// }
+///
+/// fn receiver(
+///     receiver_secret: &JubJubScalar,
+///     sender_public: &ExtendedPoint,
+///     cipher: &PoseidonCipher,
+///     nonce: &BlsScalar,
+/// ) -> [BlsScalar; MESSAGE_CAPACITY] {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(receiver_secret, sender_public);
+///
+///     // Decrypt the message
+///     cipher
+///         .decrypt(&shared_secret, &nonce)
+///         .expect("Failed to decrypt!")
+/// }
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Default)]
+pub struct PoseidonCipher {
+    cipher: [BlsScalar; CIPHER_SIZE],
+}
+
+impl PoseidonCipher {
+    /// [`PoseidonCipher`] constructor
+    pub fn new(cipher: [BlsScalar; CIPHER_SIZE]) -> Self {
+        Self { cipher }
+    }
+
+    /// Maximum number of scalars allowed per message
+    pub fn capacity() -> usize {
+        MESSAGE_CAPACITY
+    }
+
+    /// Bytes consumed on serialization of the poseidon cipher
+    pub fn serialized_size() -> usize {
+        ENCRYPTED_DATA_SIZE
+    }
+
+    /// Encrypt a slice of scalars into an internal cipher representation
+    ///
+    /// The message size will be truncated to [`MESSAGE_CAPACITY`] bits
+    pub fn encrypt(
+        message: &[BlsScalar],
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Self {
+        let zero = BlsScalar::zero();
+        let mut strategy = ScalarStrategy::new();
+
+        let mut cipher = [zero; CIPHER_SIZE];
+        let mut state = PoseidonCipher::initial_state(secret, *nonce);
+
+        strategy.perm(&mut state);
+
+        (0..MESSAGE_CAPACITY).for_each(|i| {
+            state[i + 1] += if i < message.len() {
+                message[i]
+            } else {
+                BlsScalar::zero()
+            };
+
+            cipher[i] = state[i + 1];
+        });
+
+        strategy.perm(&mut state);
+        cipher[MESSAGE_CAPACITY] = state[1];
+
+        PoseidonCipher::new(cipher)
+    }
+
+    /// Perform the decrypt of a previously encrypted message.
+    ///
+    /// Will return `None` if the decryption fails.
+    pub fn decrypt(
+        &self,
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Option<[BlsScalar; MESSAGE_CAPACITY]> {
+        let zero = BlsScalar::zero();
+        let mut strategy = ScalarStrategy::new();
+
+        let mut message = [zero; MESSAGE_CAPACITY];
+        let mut state = PoseidonCipher::initial_state(secret, *nonce);
+
+        strategy.perm(&mut state);
+
+        (0..MESSAGE_CAPACITY).for_each(|i| {
+            message[i] = self.cipher[i] - state[i + 1];
+            state[i + 1] = self.cipher[i];
+        });
+
+        strategy.perm(&mut state);
+
+        if self.cipher[MESSAGE_CAPACITY] == state[1] {
+            Some(message)
+        } else {
+            None
+        }
+    }
+
+    /// Getter for the cipher
+    pub fn cipher(&self) -> &[BlsScalar; CIPHER_SIZE] {
+        &self.cipher
+    }
```

Comment:
Why not making the `cipher` public directly on the declaration?

Reply:
We should never use the attributes directly because the structure itself is prone to change.

If we encapsulate the data access with getters, then we have more flexibility to change the structure in the future.

Reply:
Makes sense!